File: minigit-4/py/commands/cat_file.py
"""cat-file command - Examine object internals."""

import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent))

from utils import find_repo_root
from objects import read_object, Commit, Tree, Tag, resolve_object


def run(args):
    """Examine git objects."""
    repo_root = find_repo_root()
    if repo_root is None:
        print("fatal: not a minigit repository", file=sys.stderr)
        return 1

    # Parse arguments
    show_type = False
    show_size = False
    pretty = False
    obj_type = None
    target = None

    i = 0
    while i < len(args):
        arg = args[i]
        if arg == '-t':
            show_type = True
        elif arg == '-s':
            show_size = True
        elif arg == '-p':
            pretty = True
        elif arg in ('blob', 'tree', 'commit', 'tag'):
            obj_type = arg
            if i + 1 < len(args):
                target = args[i + 1]
                i += 1
        elif not arg.startswith('-'):
            target = arg
        i += 1

    if not target:
        print("fatal: no object specified", file=sys.stderr)
        return 1

    # Resolve object
    try:
        sha = resolve_object(target, repo_root)
    except ValueError as e:
        print(f"fatal: Not a valid object name {target}", file=sys.stderr)
        return 1

    try:
        actual_type, data = read_object(sha, repo_root)
    except ValueError as e:
        print(f"fatal: Not a valid object name {target}", file=sys.stderr)
        return 1

    if show_type:
        print(actual_type)
        return 0

    if show_size:
        print(len(data))
        return 0

    if obj_type:
        # Raw output of specific type
        if actual_type != obj_type:
            print(f"fatal: object {sha} is {actual_type}, not {obj_type}", file=sys.stderr)
            return 1
        sys.stdout.buffer.write(data)
        return 0

    if pretty:
        if actual_type == 'blob':
            sys.stdout.buffer.write(data)
        elif actual_type == 'tree':
            tree = Tree.deserialize(data)
            for entry in tree.entries:
                if entry.mode.startswith('40'):
                    entry_type = 'tree'
                else:
                    entry_type = 'blob'
                print(f'{entry.mode} {entry_type} {entry.sha}\t{entry.name}')
        elif actual_type == 'commit':
            commit = Commit.deserialize(data)
            print(f'tree {commit.tree_sha}')
            for parent in commit.parents:
                print(f'parent {parent}')
            print(f'author {commit.author}')
            print(f'committer {commit.committer}')
            print()
            print(commit.message)
        elif actual_type == 'tag':
            tag = Tag.deserialize(data)
            print(f'object {tag.object_sha}')
            print(f'type {tag.object_type}')
            print(f'tag {tag.tag_name}')
            print(f'tagger {tag.tagger}')
            print()
            print(tag.message)
        else:
            print(f"Unknown object type: {actual_type}", file=sys.stderr)
            return 1
        return 0

    # Default: raw output
    sys.stdout.buffer.write(data)
    return 0

File: minigit-4/ts/src/commands/cat-file.ts
// cat-file command - Examine objects

import { findRepoRoot } from '../utils';
import { readObject, parseCommitContent, parseTreeContent, parseTagContent, objectExists, resolveShortSha } from '../objects';
import { resolveRevision } from '../refs';

export function catFile(args: string[]): number {
  const repoRoot = findRepoRoot();
  if (!repoRoot) {
    console.error('fatal: not a minigit repository');
    return 1;
  }

  let showType = false;
  let showSize = false;
  let prettyPrint = false;
  let objectType: string | null = null;
  let objectRef: string | null = null;

  for (let i = 0; i < args.length; i++) {
    if (args[i] === '-t') {
      showType = true;
    } else if (args[i] === '-s') {
      showSize = true;
    } else if (args[i] === '-p') {
      prettyPrint = true;
    } else if (['blob', 'tree', 'commit', 'tag'].includes(args[i])) {
      objectType = args[i];
    } else if (!args[i].startsWith('-')) {
      objectRef = args[i];
    }
  }

  if (!objectRef) {
    console.error('fatal: object reference required');
    return 1;
  }

  // Resolve the object reference
  let sha = resolveRevision(repoRoot, objectRef);

  if (!sha) {
    // Try as short SHA
    sha = resolveShortSha(repoRoot, objectRef);
  }

  if (!sha || !objectExists(repoRoot, sha)) {
    console.error(`fatal: Not a valid object name ${objectRef}`);
    return 1;
  }

  const { type, size, content } = readObject(repoRoot, sha);

  if (showType) {
    console.log(type);
    return 0;
  }

  if (showSize) {
    console.log(size);
    return 0;
  }

  if (prettyPrint) {
    return prettyPrintObject(repoRoot, type, content);
  }

  if (objectType) {
    if (objectType !== type) {
      console.error(`fatal: expected ${objectType} but got ${type}`);
      return 1;
    }
    process.stdout.write(content);
    return 0;
  }

  // Default: show content
  process.stdout.write(content);
  return 0;
}

function prettyPrintObject(repoRoot: string, type: string, content: Buffer): number {
  switch (type) {
    case 'blob':
      process.stdout.write(content);
      return 0;

    case 'tree':
      return prettyPrintTree(content);

    case 'commit':
      return prettyPrintCommit(content);

    case 'tag':
      return prettyPrintTag(content);

    default:
      console.error(`Unknown object type: ${type}`);
      return 1;
  }
}

function prettyPrintTree(content: Buffer): number {
  const entries = parseTreeContent(content);

  for (const entry of entries) {
    const typeStr = entry.mode === '40000' ? 'tree' : 'blob';
    // Pad mode to 6 characters
    const modeStr = entry.mode.padStart(6, '0');
    console.log(`${modeStr} ${typeStr} ${entry.sha}\t${entry.name}`);
  }

  return 0;
}

function prettyPrintCommit(content: Buffer): number {
  const info = parseCommitContent(content);

  console.log(`tree ${info.tree}`);
  for (const parent of info.parents) {
    console.log(`parent ${parent}`);
  }
  console.log(`author ${info.author} <${info.authorEmail}> ${info.authorTimestamp} ${info.authorTz}`);
  console.log(`committer ${info.committer} <${info.committerEmail}> ${info.committerTimestamp} ${info.committerTz}`);
  console.log('');
  console.log(info.message);

  return 0;
}

function prettyPrintTag(content: Buffer): number {
  const info = parseTagContent(content);

  console.log(`object ${info.object}`);
  console.log(`type ${info.type}`);
  console.log(`tag ${info.tag}`);
  console.log(`tagger ${info.tagger} <${info.taggerEmail}> ${info.taggerTimestamp} ${info.taggerTz}`);
  console.log('');
  console.log(info.message);

  return 0;
}
